File: ide/src/main/java/org/dau/ui/schematic/fx/model/FxBlock.java
package org.dau.ui.schematic.fx.model;

import javafx.beans.property.SimpleDoubleProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.dau.runtime.Block;
import org.dau.util.Xmls;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.lang.invoke.MethodHandles;
import java.lang.reflect.*;
import java.util.ArrayList;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.stream.Stream;

import static java.lang.invoke.MethodType.fromMethodDescriptorString;
import static java.util.Arrays.stream;
import static java.util.Comparator.comparingInt;

public final class FxBlock {

  public final FxSchema schema;
  public final int id;

  public final SimpleStringProperty name = new SimpleStringProperty(this, "name");
  public final SimpleObjectProperty<Executable> executable = new SimpleObjectProperty<>(this, "executable");

  public final SimpleDoubleProperty x = new SimpleDoubleProperty(this, "x");
  public final SimpleDoubleProperty y = new SimpleDoubleProperty(this, "y");
  public final SimpleDoubleProperty w = new SimpleDoubleProperty(this, "w");
  public final SimpleDoubleProperty h = new SimpleDoubleProperty(this, "h");

  public final ObservableList<Input> inputs = FXCollections.observableArrayList();
  public final ObservableList<Output> outputs = FXCollections.observableArrayList();

  public FxBlock(FxSchema schema, Executable executable) {
    this(schema, schema.blockId(), executable);
  }

  private FxBlock(FxSchema schema, int id, Executable executable) {
    this.schema = schema;
    this.id = id;
    collectInputs();
    collectOutputs();
    this.executable.set(executable);
    this.name.set(getMetaName() + " " + id);
  }

  private void collectInputs() {
    executable.addListener((o, ov, e) -> {
      var params = e.getParameters();
      var inputs = stream(params).map(p -> new Input(p.getName())).toList();
      this.inputs.removeIf(i -> inputs.stream().noneMatch(i::equals));
      inputs.forEach(input -> {
        if (this.inputs.stream().noneMatch(input::equals)) {
          this.inputs.add(input);
        }
      });
      this.inputs.sort(comparingInt(inputs::indexOf));
    });
  }

  private void collectOutputs(Type type, ArrayList<Output> outputs) {
    if (type != void.class) {
      outputs.add(new Output("@"));
    }
    if (type instanceof ParameterizedType t) {
      collectOutputs(t.getRawType(), outputs);
    } else if (type instanceof Class<?> c) {
      for (var m : c.getMethods()) {
        if (m.getParameterCount() > 2) {
          continue;
        }
        if (m.isSynthetic() || m.isBridge() || m.isVarArgs()) {
          continue;
        }
        if (m.getDeclaringClass() == Object.class) {
          continue;
        }
        outputs.add(new Output(m.getName()));
      }
    }
  }

  private void collectOutputs() {
    executable.addListener((o, ov, e) -> {
      var outputs = new ArrayList<Output>();
      collectOutputs(e.getAnnotatedReturnType().getType(), outputs);
      this.outputs.removeIf(out -> outputs.stream().noneMatch(out::equals));
      outputs.forEach(output -> {
        if (this.outputs.stream().noneMatch(output::equals)) {
          this.outputs.add(output);
        }
      });
      this.outputs.sort(comparingInt(outputs::indexOf));
    });
  }

  public String getVar() {
    return "v_" + id;
  }

  public String getMetaName() {
    var executable = this.executable.get();
    var block = executable.getAnnotation(Block.class);
    if (block != null) {
      return block.value();
    }
    block = executable.getDeclaringClass().getAnnotation(Block.class);
    if (block != null) {
      return block.value();
    }
    return executable.getName();
  }

  public Input getInput(String id) {
    return inputs.parallelStream()
      .filter(i -> i.id.equals(id))
      .findAny()
      .orElse(null);
  }

  public Output getOutput(String id) {
    return outputs.parallelStream()
      .filter(o -> o.id.equals(id))
      .findAny()
      .orElse(null);
  }

  public Stream<Input> getInputs() {
    return inputs.stream();
  }

  public Stream<Output> getOutputs() {
    return outputs.stream();
  }

  public void remove() {
    if (!schema.removeBlock(this)) {
      throw new NoSuchElementException(toString());
    }
  }

  public Element toXml(Document doc) {
    var el = doc.createElement("block");
    el.setAttribute("id", Integer.toString(id));
    el.setAttribute("name", name.get());
    el.setAttribute("x", Double.toString(x.get()));
    el.setAttribute("y", Double.toString(y.get()));
    el.setAttribute("w", Double.toString(w.get()));
    el.setAttribute("h", Double.toString(h.get()));
    try {
      var lookup = MethodHandles.publicLookup();
      if (executable.get() instanceof Constructor<?> c) {
        var mh = lookup.unreflectConstructor(c);
        el.setAttribute("type", mh.type().descriptorString());
        el.setAttribute("executable", "*");
      } else if (executable.get() instanceof Method m) {
        var mh = lookup.unreflect(m);
        el.setAttribute("class", m.getDeclaringClass().getName());
        el.setAttribute("type", mh.type().descriptorString());
        el.setAttribute("executable", m.getName());
      }
    } catch (IllegalAccessException e) {
      throw new IllegalStateException(e);
    }
    var inputsEl = doc.createElement("inputs");
    inputs.forEach(input -> {
      var c = input.constant.get();
      if (c != null) {
        var e = doc.createElement("input");
        e.setAttribute("id", input.id);
        e.setTextContent(c);
        inputsEl.appendChild(e);
      }
    });
    if (inputsEl.hasChildNodes()) {
      el.appendChild(inputsEl);
    }
    return el;
  }

  public static FxBlock fromXml(FxSchema schema, ClassLoader classLoader, Element element) {
    int id = Integer.parseInt(element.getAttribute("id"));
    var executableName = element.getAttribute("executable");
    final Executable executable;
    try {
      var mt = fromMethodDescriptorString(element.getAttribute("type"), classLoader);
      if (executableName.equals("*")) {
        executable = mt.returnType().getConstructor(mt.parameterArray());
      } else {
        var className = element.getAttribute("class");
        var cl = Class.forName(className, false, classLoader);
        executable = cl.getMethod(executableName, mt.parameterArray());
      }
    } catch (NoSuchMethodException | ClassNotFoundException e) {
      throw new IllegalStateException(e);
    }
    var block = new FxBlock(schema, id, executable);
    block.name.set(element.getAttribute("name"));
    block.x.set(Double.parseDouble(element.getAttribute("x")));
    block.y.set(Double.parseDouble(element.getAttribute("y")));
    block.w.set(Double.parseDouble(element.getAttribute("w")));
    block.h.set(Double.parseDouble(element.getAttribute("h")));
    Xmls.elementsByTag(element, "inputs", "input").forEach(e -> {
      var inputId = e.getAttribute("id");
      var input = block.getInput(inputId);
      if (input != null) {
        input.constant.set(e.getTextContent());
      }
    });
    return block;
  }

  @Override
  public int hashCode() {
    return id;
  }

  @Override
  public boolean equals(Object o) {
    return o instanceof FxBlock b && id == b.id;
  }

  @Override
  public String toString() {
    return "Block(" + id + ")";
  }

  public final class Input {

    public final String id;
    public final SimpleStringProperty constant = new SimpleStringProperty(this, "constant");

    public Input(String id) {
      this.id = id;
    }

    public FxBlock getBlock() {
      return FxBlock.this;
    }

    @Override
    public int hashCode() {
      return Objects.hash(id, getBlock().id);
    }

    @Override
    public boolean equals(Object o) {
      return o instanceof Input b && b.id.equals(id) && b.getBlock().equals(getBlock());
    }

    @Override
    public String toString() {
      return FxBlock.this + "@in:" + id;
    }
  }

  public final class Output {

    public final String id;

    public Output(String id) {
      this.id = id;
    }

    public FxBlock getBlock() {
      return FxBlock.this;
    }

    @Override
    public int hashCode() {
      return Objects.hash(id, getBlock().id);
    }

    @Override
    public boolean equals(Object o) {
      return o instanceof Output b && id.equals(b.id) && b.getBlock().equals(getBlock());
    }

    @Override
    public String toString() {
      return FxBlock.this + "@out:" + id;
    }
  }
}

File: ide/src/main/kotlin/org/dau/ui/schematic/model/FxBlock.kt
package org.dau.ui.schematic.model

import javafx.beans.property.SimpleDoubleProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import org.dau.runtime.Block
import org.dau.ui.utils.Xmls
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType.fromMethodDescriptorString
import java.lang.reflect.*
import java.util.*
import java.util.Arrays.stream
import java.util.Comparator.comparingInt
import java.util.stream.Stream

class FxBlock private constructor(val schema: FxSchema, val id: Int, executable: Executable) {

  val name = SimpleStringProperty(this, "name")
  val executable = SimpleObjectProperty<Executable>(this, "executable")

  val x = SimpleDoubleProperty(this, "x")
  val y = SimpleDoubleProperty(this, "y")
  val w = SimpleDoubleProperty(this, "w")
  val h = SimpleDoubleProperty(this, "h")

  val inputs = FXCollections.observableArrayList<Input>()
  val outputs = FXCollections.observableArrayList<Output>()

  val varName: String
    get() = "v_$id"

  val metaName: String
    get() {
      val executable = this.executable.get()
      var block: Block? = executable.getAnnotation(Block::class.java)
      if (block != null) {
        return block.value
      }
      block = executable.declaringClass.getAnnotation(Block::class.java)
      return block?.value ?: executable.name
    }

  constructor(schema: FxSchema, executable: Executable) : this(schema, schema.blockId(), executable) {}

  init {
    collectInputs()
    collectOutputs()
    this.executable.set(executable)
    this.name.set("$metaName $id")
  }

  private fun collectInputs() {
    executable.addListener { _, _, e ->
      val params = e.parameters
      val inputs = stream(params).map { p -> Input(p.name) }.toList()
      this.inputs.removeIf { i -> inputs.stream().noneMatch { i == it } }
      inputs.forEach { input ->
        if (this.inputs.stream().noneMatch { input == it }) {
          this.inputs.add(input)
        }
      }
      this.inputs.sortWith(comparingInt { inputs.indexOf(it) })
    }
  }

  private fun collectOutputs(type: Type, outputs: ArrayList<Output>) {
    if (type !== Void.TYPE) {
      outputs.add(Output("@"))
    }
    if (type is ParameterizedType) {
      collectOutputs(type.rawType, outputs)
    } else if (type is Class<*>) {
      for (m in type.methods) {
        if (m.parameterCount > 2) {
          continue
        }
        if (m.isSynthetic || m.isBridge || m.isVarArgs) {
          continue
        }
        if (m.declaringClass == Any::class.java) {
          continue
        }
        outputs.add(Output(m.name))
      }
    }
  }

  private fun collectOutputs() {
    executable.addListener { _, _, e ->
      val outputs = ArrayList<Output>()
      collectOutputs(e.annotatedReturnType.type, outputs)
      this.outputs.removeIf { out -> outputs.stream().noneMatch { out == it } }
      outputs.forEach { output ->
        if (this.outputs.stream().noneMatch { output == it }) {
          this.outputs.add(output)
        }
      }
      this.outputs.sortWith(comparingInt { outputs.indexOf(it) })
    }
  }

  fun getInput(id: String): Input? {
    return inputs.parallelStream()
      .filter { i -> i.id == id }
      .findAny()
      .orElse(null)
  }

  fun getOutput(id: String): Output? {
    return outputs.parallelStream()
      .filter { o -> o.id == id }
      .findAny()
      .orElse(null)
  }

  fun getInputs(): Stream<Input> {
    return inputs.stream()
  }

  fun getOutputs(): Stream<Output> {
    return outputs.stream()
  }

  fun remove() {
    if (!schema.removeBlock(this)) {
      throw NoSuchElementException(toString())
    }
  }

  fun toXml(doc: Document): Element {
    val el = doc.createElement("block")
    el.setAttribute("id", id.toString())
    el.setAttribute("name", name.get())
    el.setAttribute("x", x.get().toString())
    el.setAttribute("y", y.get().toString())
    el.setAttribute("w", w.get().toString())
    el.setAttribute("h", h.get().toString())
    try {
      val lookup = MethodHandles.publicLookup()
      when (val e = executable.get()) {
        is Constructor<*> -> {
          val mh = lookup.unreflectConstructor(e)
          el.setAttribute("type", mh.type().descriptorString())
          el.setAttribute("executable", "*")
        }
        is Method -> {
          val mh = lookup.unreflect(e)
          el.setAttribute("class", e.getDeclaringClass().name)
          el.setAttribute("type", mh.type().descriptorString())
          el.setAttribute("executable", e.name)
        }
      }
    } catch (e: IllegalAccessException) {
      throw IllegalStateException(e)
    }

    val inputsEl = doc.createElement("inputs")
    inputs.forEach { input ->
      val c = input.constant.get()
      if (c != null) {
        val e = doc.createElement("input")
        e.setAttribute("id", input.id)
        e.textContent = c
        inputsEl.appendChild(e)
      }
    }
    if (inputsEl.hasChildNodes()) {
      el.appendChild(inputsEl)
    }
    return el
  }

  override fun hashCode(): Int {
    return id
  }

  override fun equals(other: Any?): Boolean {
    return other is FxBlock && id == other.id
  }

  override fun toString(): String {
    return "Block($id)"
  }

  inner class Input(val id: String) {
    val constant = SimpleStringProperty(this, "constant")

    val block: FxBlock
      get() = this@FxBlock

    override fun hashCode(): Int {
      return Objects.hash(id, block.id)
    }

    override fun equals(other: Any?): Boolean {
      return other is Input && other.id == id && other.block == block
    }

    override fun toString(): String {
      return this@FxBlock.toString() + "@in:" + id
    }
  }

  inner class Output(val id: String) {

    val block: FxBlock
      get() = this@FxBlock

    override fun hashCode(): Int {
      return Objects.hash(id, block.id)
    }

    override fun equals(other: Any?): Boolean {
      return other is Output && id == other.id && other.block == block
    }

    override fun toString(): String {
      return this@FxBlock.toString() + "@out:" + id
    }
  }

  companion object {

    fun fromXml(schema: FxSchema, classLoader: ClassLoader, element: Element): FxBlock {
      val id = Integer.parseInt(element.getAttribute("id"))
      val executableName = element.getAttribute("executable")
      val executable = try {
        val mt = fromMethodDescriptorString(element.getAttribute("type"), classLoader)
        if (executableName == "*") {
          mt.returnType().getConstructor(*mt.parameterArray())
        } else {
          val className = element.getAttribute("class")
          val cl = Class.forName(className, false, classLoader)
          cl.getMethod(executableName, *mt.parameterArray())
        }
      } catch (e: NoSuchMethodException) {
        throw IllegalStateException(e)
      } catch (e: ClassNotFoundException) {
        throw IllegalStateException(e)
      }

      val block = FxBlock(schema, id, executable)
      block.name.set(element.getAttribute("name"))
      block.x.set(java.lang.Double.parseDouble(element.getAttribute("x")))
      block.y.set(java.lang.Double.parseDouble(element.getAttribute("y")))
      block.w.set(java.lang.Double.parseDouble(element.getAttribute("w")))
      block.h.set(java.lang.Double.parseDouble(element.getAttribute("h")))
      Xmls.elementsByTag(element, "inputs", "input").forEach { e ->
        val inputId = e.getAttribute("id")
        val input = block.getInput(inputId)
        input?.constant?.set(e.textContent)
      }
      return block
    }
  }
}
